Remove CustomComponentBase and use Reflection instead

SQL-MisterMagoo opened the pull request.

egil commented on src/BlazorWebFormsComponents/BaseWebFormsComponent.cs:
```diff
@@ -226,21 +226,27 @@ public ValueTask DisposeAsync()
+			_renderFragmentField.SetValue(this, (RenderFragment)(builder =>
 				{
 					builder.OpenComponent(1, typeof(CascadingValue<BaseWebFormsComponent>));
 					builder.AddAttribute(2, "Name", "ParentComponent");
```
To make things more refactor friendly, I suggest changing the Parent property to this:

```csharp
		[CascadingParameter(Name=nameof(Parent))]
		public virtual BaseWebFormsComponent Parent { get; set; }
```

Then the follow could be:

```suggestion
					builder.AddAttribute(2, "Name", nameof(Parent));
					builder.AddAttribute(5, "Fixed", true); // This should be possible here as well
```

SQL-MisterMagoo replied:
Yes, good catch!

egil replied:
Actually, you could even remove the other magic strings also:

```csharp
builder.AddAttribute(2, nameof(CascadingValue<BaseWebFormsComponent>.Name), nameof(Parent));
```

egil commented on src/BlazorWebFormsComponents/BaseWebFormsComponent.cs:
```diff
@@ -226,21 +226,27 @@ public ValueTask DisposeAsync()
 
 		public BaseWebFormsComponent ()
 		{
+			// Get Access to the ComponentBase fields we need to wrap every component in a CascadingValue
```
To speed things up a little, the result of these reflection api calls can be cached in static readonly fields, so they are only performed once.

SQL-MisterMagoo replied:
True - I totally agree.

egil commented on src/BlazorWebFormsComponents/BaseWebFormsComponent.cs:
```diff
@@ -226,21 +226,27 @@ public ValueTask DisposeAsync()
 			RenderFragment _standardRender = builder2 => {
-				_hasPendingQueuedRender = false;
-				_hasNeverRendered = false;
+				_hasPendingQueuedRenderField.SetValue(this,false);
```
Curious, why is setting `_hasPendingQueuedRenderField` and `_hasNeverRenderedField` necessary?

SQL-MisterMagoo replied:
It's part of the standard renderFragment from ComponentBase - I don't see any reason to change the standard.

SQL-MisterMagoo replied:
https://github.com/dotnet/aspnetcore/blob/1e735afa929e958dd62a7e68db990bbacd675a25/src/Components/Components/src/ComponentBase.cs#L40

egil replied:
I see. Why do you need to set them in your renderFragment action if they are set by the `ComponentBase._renderFragment`? When the rendering has been started, which it has at this point, then it should not matter?

egil replied:
Ahh nevermind, you are not reusing the _renderFragment, you are replacing it. Yep, it looks right :)

egil replied:
OK, but maybe you can minimize required reflection of the base code, and just grab the renderFragment and call that, e.g.:

```csharp
var _renderFragmentField = typeof(ComponentBase).GetField("_renderFragment", BindingFlags.NonPublic | BindingFlags.Instance);

var baseRenderFragment = _renderFragmentField.GetValue(this);

// Override the default RenderFragment with our Special Sauce version
_renderFragmentField.SetValue(this, (RenderFragment)(builder =>
{
	builder.OpenComponent(1, typeof(CascadingValue<BaseWebFormsComponent>));
	builder.AddAttribute(2, nameof(CascadingValue<BaseWebFormsComponent>.Name), nameof(ParentComponent));
	builder.AddAttribute(3, nameof(CascadingValue<BaseWebFormsComponent>.Value), this);
	builder.AddAttribute(4, nameof(CascadingValue<BaseWebFormsComponent>.Fixed), true);
	builder.AddAttribute(5, nameof(CascadingValue<BaseWebFormsComponent>.ChildContent), baseRenderFragment);
	builder.CloseComponent();
}));
```

SQL-MisterMagoo replied:
Yes, I'll try that out and see - it would certainly be better to just use the default rather than recreating it.

egil replied:
ps. i like how reflection allows us to completely ignore the fact that the field _renderFragment is readonly :)

SQL-MisterMagoo replied:
I'm not 100% sure it does - we are, after all, setting it in the constructor, which is allowed.

I haven't tried setting it elsewhere, but I would hope it would fail.

egil replied:
Ahh yeah, lets hope thats the case.

SQL-MisterMagoo commented:
@csharpfritz 
Hey Jeff, don't merge this yet - if I could mark it as draft now i would - I like the suggested changes from @egil and will update the PR with them

SQL-MisterMagoo commented:
Ok, I'm happy with this - it uses the ComponentBase default RenderFragment now instead of replacing it, and I removed the magic strings, cached the reflected FieldInfo and modified the comments, so future devs know what we were trying to achieve.

egil commented on src/BlazorWebFormsComponents/BaseWebFormsComponent.cs:
```diff
@@ -223,24 +212,29 @@ public ValueTask DisposeAsync()
 		#endregion
 
 		public bool LayoutTemplateRendered { get; set; } = false;
+		private static FieldInfo _renderFragmentField;
```
```suggestion
		private static readonly FieldInfo _renderFragmentField = typeof(ComponentBase).GetField(BASEFRAGMENTFIELDNAME, BindingFlags.NonPublic | BindingFlags.Instance);
```

egil commented on src/BlazorWebFormsComponents/BaseWebFormsComponent.cs:
```diff
@@ -223,24 +212,29 @@ public ValueTask DisposeAsync()
-			};
-			_renderFragment = builder =>
+			// Get Access to the ComponentBase field we need to wrap every component in a CascadingValue
+			_renderFragmentField ??= typeof(ComponentBase).GetField(BASEFRAGMENTFIELDNAME, BindingFlags.NonPublic | BindingFlags.Instance);
```
```suggestion
```

egil commented on src/BlazorWebFormsComponents/BaseWebFormsComponent.cs:
```diff
@@ -223,24 +212,29 @@ public ValueTask DisposeAsync()
+		private const string PARENTCOMPONENTNAME = "ParentComponent";
+		//if you added ChildContent to this class, we could just nameof(ChildContent)
+		private const string CHILDCONTENTNAME = "ChildContent";
 
```
```suggestion
private readonly RenderFragment _baseRenderFragment;
```

egil commented on src/BlazorWebFormsComponents/BaseWebFormsComponent.cs:
```diff
@@ -223,24 +212,29 @@ public ValueTask DisposeAsync()
+			_renderFragmentField ??= typeof(ComponentBase).GetField(BASEFRAGMENTFIELDNAME, BindingFlags.NonPublic | BindingFlags.Instance);
+
+			// Grab a copy of the default RenderFragment to go into the CascadingValue
+			var _standardRender = (RenderFragment)_renderFragmentField.GetValue(this);
```
```suggestion
			_baseRenderFragment = (RenderFragment)_renderFragmentField.GetValue(this);

```

egil commented on src/BlazorWebFormsComponents/BaseWebFormsComponent.cs:
```diff
@@ -223,24 +212,29 @@ public ValueTask DisposeAsync()
-				};
+				}));
 		}
 
```
This makes it more obvious what is going on I think: 
```suggestion
private void ParentWrappingBuildRenderTree(RenderTreeBuilder builder)
{
    builder.OpenComponent(1, typeof(CascadingValue<BaseWebFormsComponent>));
    builder.AddAttribute(2, nameof(CascadingValue<object>.Name), PARENTCOMPONENTNAME);
    builder.AddAttribute(3, nameof(CascadingValue<object>.Value), this);
    builder.AddAttribute(4, CHILDCONTENTNAME, _baseRenderFragment);
}
```

egil commented on src/BlazorWebFormsComponents/BaseWebFormsComponent.cs:
```diff
@@ -223,24 +212,29 @@ public ValueTask DisposeAsync()
+			var _standardRender = (RenderFragment)_renderFragmentField.GetValue(this);
+
+			// Override the default RenderFragment with our Special Sauce version
+			_renderFragmentField.SetValue(this, (RenderFragment)(builder =>
```
```suggestion
			_renderFragmentField.SetValue(this, ParentWrappingBuildRenderTree);
```

egil commented on src/BlazorWebFormsComponents/BaseWebFormsComponent.cs:
```diff
@@ -223,24 +212,29 @@ public ValueTask DisposeAsync()
+					builder.AddAttribute(4, CHILDCONTENTNAME, _standardRender);
 					builder.CloseComponent();
-				};
+				}));
```
delete

egil commented:
So one more suggestion. Create method with a name that suggest what it does, and use that instead of a anonymous method created each time in the ctor. My suggested edits have been added (not tested though). 

Last time, I swear :)

ps. Why not mark the cascading value as `Fixed`?

egil commented on src/BlazorWebFormsComponents/BaseWebFormsComponent.cs:
```diff
@@ -223,24 +212,29 @@ public ValueTask DisposeAsync()
+		private const string BASEFRAGMENTFIELDNAME = "_renderFragment";
+		private const string PARENTCOMPONENTNAME = "ParentComponent";
+		//if you added ChildContent to this class, we could just nameof(ChildContent)
+		private const string CHILDCONTENTNAME = "ChildContent";
```
Not needed.
```suggestion
```

SQL-MisterMagoo commented:
It seems those changes from @egil have left me failing tests - once VS finishes updating I'll fix it

SQL-MisterMagoo commented:
> @SQL-MisterMagoo so one more suggestion. Create method with a name that suggest what it does (ok naming is hard, so maybe `ParentWrappingBuildRenderTree` is not the best), and use that instead of a anonymous method created each time in the ctor. My suggested edits have been added (not tested though).
> 
> Last time, I swear :)
> 
> ps. Why not mark the cascading value as `Fixed`?

I have added `IsFixed` to the CascadingValue for the performance savings and sorted out the errors.
I made the new method a local method as nothing else needs to have access to it.

egil commented:
> I made the new method a local method as nothing else needs to have access to it.

I think that is a good use of local methods actually. When private isnt enough, make them local!

However, you might also want to move the _baseRenderFragment variable into the ctor again then, since that is only used by the local method, and will be captured by it.

SQL-MisterMagoo commented:
> However, you might also want to move the _baseRenderFragment variable into the ctor again then, since that is only used by the local method, and will be captured by it.

[You swore that was the last!](https://github.com/FritzAndFriends/BlazorWebFormsComponents/pull/118#issuecomment-597782467)

You are possibly right, but I haven't got the will to change it now 😃

egil commented:
> [You swore that was the last!](https://github.com/FritzAndFriends/BlazorWebFormsComponents/pull/118#issuecomment-597782467)

Well, I did so with a smiley at the end, so doesn't count ;)

> You are possibly right, but I haven't got the will to change it now 😃

Hehe fair enough, I will get of your back for now ;)
